File: src/runflow/core/graph.py
from __future__ import annotations

from dataclasses import dataclass

from runflow.core.node import Node
from runflow.core.types import dtype_accepts


@dataclass(frozen=True)
class Endpoint:
    node_id: str
    port: str


@dataclass(frozen=True)
class Edge:
    source: Endpoint
    target: Endpoint


class Graph:
    def __init__(self) -> None:
        self.nodes: dict[str, Node] = {}
        self.edges: list[Edge] = []

    def add_node(self, node: Node) -> Node:
        if node.id in self.nodes:
            raise ValueError(f"Duplicate node id: {node.id}")
        self.nodes[node.id] = node
        return node

    def connect(self, source_node: str, source_port: str, target_node: str, target_port: str) -> None:
        if source_node not in self.nodes:
            raise KeyError(f"Unknown source node: {source_node}")
        if target_node not in self.nodes:
            raise KeyError(f"Unknown target node: {target_node}")

        source_ports = self.nodes[source_node].OUTPUTS
        target_ports = self.nodes[target_node].INPUTS

        if source_port not in source_ports:
            raise KeyError(f"Node {source_node} has no output port {source_port}")
        if target_port not in target_ports:
            raise KeyError(f"Node {target_node} has no input port {target_port}")

        src = source_ports[source_port]
        dst = target_ports[target_port]
        if not dtype_accepts(dst.dtype, src.dtype):
            raise TypeError(
                f"Cannot connect {source_node}.{source_port} ({src.dtype.name}) "
                f"to {target_node}.{target_port} ({dst.dtype.name})"
            )

        self.edges.append(Edge(Endpoint(source_node, source_port), Endpoint(target_node, target_port)))

    def incoming_edges(self, node_id: str) -> list[Edge]:
        return [edge for edge in self.edges if edge.target.node_id == node_id]

    def outgoing_edges(self, node_id: str) -> list[Edge]:
        return [edge for edge in self.edges if edge.source.node_id == node_id]

    def source_nodes(self) -> list[Node]:
        targets = {edge.target.node_id for edge in self.edges}
        return [node for node_id, node in self.nodes.items() if node_id not in targets]

    def input_nodes(self) -> list[Node]:
        return [node for node in self.nodes.values() if node.IS_INPUT]

    def sink_nodes(self) -> list[Node]:
        sources = {edge.source.node_id for edge in self.edges}
        return [node for node_id, node in self.nodes.items() if node_id not in sources]

File: src/frontend/src/features/workflows/graph.ts
import type { SchemaValues } from "@/shared/schema-form/types";
import type { WorkflowEdge, WorkflowGraph, WorkflowNode, WorkflowSchema } from "./types";

export function typeAccepts(schema: WorkflowSchema, targetType: string, sourceType: string): boolean {
  if (!schema.types[targetType] || !schema.types[sourceType]) {
    throw new Error(`Unknown port type: ${targetType} or ${sourceType}`);
  }
  return targetType === sourceType;
}

export function nodeAccent(schema: WorkflowSchema, nodeType: string): string {
  const info = schema.nodes[nodeType];
  if (!info) throw new Error(`Unknown node type: ${nodeType}`);
  const output = Object.values(info.outputs)[0];
  if (!output) return "#3a4353";
  const type = schema.types[output.type];
  if (!type) throw new Error(`Unknown output type: ${output.type}`);
  return type.color;
}

export function addNode(schema: WorkflowSchema, graph: WorkflowGraph, type: string, x: number, y: number): WorkflowGraph {
  const info = schema.nodes[type];
  if (!info) throw new Error(`Unknown node type: ${type}`);
  const count = graph.nodes.filter((node) => node.type === type).length + 1;
  const node: WorkflowNode = {
    id: `${type}_${count}`,
    type,
    x,
    y,
    params: structuredClone(info.settings_defaults) as SchemaValues,
    runtime: structuredClone(info.runtime_defaults) as SchemaValues,
  };
  return { ...graph, nodes: [...graph.nodes, node] };
}

export function deleteNodes(graph: WorkflowGraph, nodeIds: string[]): WorkflowGraph {
  const removed = new Set(nodeIds);
  return {
    ...graph,
    nodes: graph.nodes.filter((node) => !removed.has(node.id)),
    edges: graph.edges.filter((edge) => !removed.has(edge.source_node) && !removed.has(edge.target_node)),
    panels: (graph.panels ?? []).map((panel) => ({
      ...panel,
      controls: panel.controls
        .map((control) => ({ ...control, targets: control.targets.filter((target) => !removed.has(target.node_id)) }))
        .filter((control) => control.targets.length > 0),
    })),
  };
}

export function renameNode(graph: WorkflowGraph, previous: string, nextId: string): WorkflowGraph {
  return {
    ...graph,
    nodes: graph.nodes.map((node) => (node.id === previous ? { ...node, id: nextId } : node)),
    edges: graph.edges.map((edge) => ({
      ...edge,
      source_node: edge.source_node === previous ? nextId : edge.source_node,
      target_node: edge.target_node === previous ? nextId : edge.target_node,
    })),
    panels: (graph.panels ?? []).map((panel) => ({
      ...panel,
      controls: panel.controls.map((control) => ({
        ...control,
        targets: control.targets.map((target) => (target.node_id === previous ? { ...target, node_id: nextId } : target)),
      })),
    })),
  };
}

export function connect(schema: WorkflowSchema, graph: WorkflowGraph, edge: WorkflowEdge): WorkflowGraph {
  const source = graph.nodes.find((node) => node.id === edge.source_node);
  const target = graph.nodes.find((node) => node.id === edge.target_node);
  if (!source || !target || source.id === target.id) return graph;
  const sourceInfo = schema.nodes[source.type];
  const targetInfo = schema.nodes[target.type];
  if (!sourceInfo || !targetInfo) throw new Error("Graph contains unknown node type");
  const sourcePort = sourceInfo.outputs[edge.source_port];
  const targetPort = targetInfo.inputs[edge.target_port];
  if (!sourcePort || !targetPort) throw new Error("Graph contains unknown port");
  if (!typeAccepts(schema, targetPort.type, sourcePort.type)) return graph;
  const duplicate = graph.edges.some((item) =>
    item.source_node === edge.source_node && item.source_port === edge.source_port
    && item.target_node === edge.target_node && item.target_port === edge.target_port,
  );
  return duplicate ? graph : { ...graph, edges: [...graph.edges, edge] };
}

export function moveNodes(graph: WorkflowGraph, nodeIds: string[], dx: number, dy: number): WorkflowGraph {
  const selected = new Set(nodeIds);
  return {
    ...graph,
    nodes: graph.nodes.map((node) => (selected.has(node.id) ? { ...node, x: node.x + dx, y: node.y + dy } : node)),
  };
}
